app/code/Magento/Catalog/Model/Product/Compare/AddToList.php

```diff
@@ -0,0 +1,154 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\Catalog\Model\Product\Compare;
+
+use Magento\Catalog\Model\CompareList;
+use Magento\Catalog\Model\CompareListFactory;
+use Magento\Catalog\Model\ResourceModel\CompareList as CompareListResource;
+use Magento\Customer\Model\Session;
+use Magento\Customer\Model\Visitor;
+
+class AddToList
+{
+    /**
+     * @var CompareListFactory
+     */
+    private $compareListFactory;
+
+    /**
+     * @var CompareList
+     */
+    private $compareList;
+
+    /**
+     * @var CompareListResource
+     */
+    private $compareListResource;
+
+    /**
+     * Customer session
+     *
+     * @var Session
+     */
+    private $customerSession;
+
+    /**
+     * Customer visitor
+     *
+     * @var Visitor
+     */
+    private $customerVisitor;
+
+    /**
+     * @param CompareListFactory $compareListFactory
+     * @param CompareList $compareList
+     * @param CompareListResource $compareListResource
+     * @param Session $customerSession
+     * @param Visitor $customerVisitor
+     */
+    public function __construct(
+        CompareListFactory $compareListFactory,
+        CompareList $compareList,
+        CompareListResource $compareListResource,
+        Session $customerSession,
+        Visitor $customerVisitor
+    ) {
+        $this->compareListFactory = $compareListFactory;
+        $this->compareList = $compareList;
+        $this->compareListResource = $compareListResource;
+        $this->customerSession = $customerSession;
+        $this->customerVisitor = $customerVisitor;
+    }
+
+    /**
+     * Get list_id
+     *
+     * @return int
+     */
+    public function execute()
+    {
+        if ($this->customerSession->isLoggedIn()) {
+            return $this->getListIdByCustomerId();
+        }
+
+        return $this->getListIdByVisitorId();
+    }
+
+    /**
+     * Set customer from visitor
+     */
+    public function setCustomerFromVisitor()
+    {
+        $customerId = $this->customerSession->getCustomerId();
```

Comment:
I would rather try to pass this information from the invocation place. Maybe we may get this data in the similar way

https://github.com/magento/magento2/blob/2fe99e93914d4b2c2a548c97ef34dca391f64ddb/app/code/Magento/Catalog/Model/ResourceModel/Product/Compare/Item.php#L216

So we don't inject the session over and over again to the service classes.

Reply:
**Done**